File: src/main/java/com/rtomyj/yugiohAPI/model/product/ProductContent.java
package com.rtomyj.yugiohAPI.model.product;

import java.util.List;
import java.util.Set;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.rtomyj.yugiohAPI.model.card.Card;

import lombok.*;
import org.springframework.hateoas.RepresentationModel;

@Data
@EqualsAndHashCode(callSuper=false)
@ToString
@Builder
@JsonInclude(JsonInclude.Include.NON_EMPTY)	// serializes non null fields - ie returns non null fields from REST request
public class ProductContent extends RepresentationModel<ProductContent>
{
	private Card card;
	private String position;
	private Set<String> rarities;



	public void setLinks()
	{
		this.card.setLinks();
	}


	public static void setLinks(@NonNull final List<ProductContent> packContents)
	{
		packContents
			.stream()
			.forEach(packContent -> packContent.setLinks());
	}
}

File: src/main/kotlin/com/rtomyj/skc/browse/product/model/ProductContent.kt
package com.rtomyj.skc.browse.product.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.rtomyj.skc.constant.SwaggerConstants
import com.rtomyj.skc.browse.card.model.Card
import io.swagger.v3.oas.annotations.media.Schema
import org.springframework.hateoas.RepresentationModel
import java.util.function.Consumer

@JsonInclude(
    JsonInclude.Include.NON_EMPTY
)
@Schema(
    implementation = Products::class,
)
data class ProductContent(
	@Schema(
        implementation = Card::class,
        description = "Information about card."
    )
    val card: Card?,
	@Schema(
        implementation = String::class,
        description = SwaggerConstants.CARD_POSITION_IN_PRODUCT_DESCRIPTION
    )
    val productPosition: String,
	@Schema(
        implementation = Set::class,
        description = SwaggerConstants.CARD_RARITIES_FOR_POSITION_DESCRIPTION
    )
    val rarities: Set<String>
) : RepresentationModel<ProductContent?>() {

    companion object {
        fun setLinks(packContents: List<ProductContent>) {
            packContents
                .forEach(Consumer { obj: ProductContent -> obj.setLinks() })
        }
    }
    

    fun setLinks() {
        card?.setLinks()
    }
}